### PR TITLE
Catalog ProductList helper has coupling to Registry #27119

### LINKED ISSUE #27119: Catalog ProductList helper has coupling to Registry
### Preconditions (*)
1. Although Helpers should be completely removed, the ones that are still in Magento core should be stateless. 

### Steps to reproduce (*)
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Helper/Product/ProductList.php#L110-L121

### Expected result (*)
1. The helper should not depend on the current category
2. The method should be moved to other, more appropriate place.

### Actual result (*)
1. Helper uses Core Registry to fetch current category and it's details

### PR DESCRIPTION
**Pull request by mcspronko**
### Description (*)
In this pull request the **Magento\Framework\Registry** class dependency is removed from the **Magento\Catalog\Helper\Product\ProductList** class.

The new **Magento\Catalog\Model\Category\Toolbar\Config** class has been added to provide access to the **Sort By** (Order Field) value. 

The **getOrderField()** method checks if the current category has the **default_sort_by** value, and returns it, the default configuration value is returned otherwise.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#27119: Catalog ProductList helper has coupling to Registry

### Manual testing scenarios (*)
1. Create a Category and add two Simple Products into the Category
2. Clear Cache
3. Verify that the Category opens correctly and uses the default **Sort By: Position** setting

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by rogyar** [comment]
Hi @mcspronko. Thank you for your change, it's really cool. Unfortunately, I have not very good news here. After the internal discussion, the conclusion was the following.
This PR will not be accepted by the product team since there's a code freeze for 2.4 and we cannot make backward-incompatible changes (such as changing signatures of public methods and removing public methods from classes marked as `@api`). 
The changeset for the next MAJOR version for the Catalog module is already defined and is not supposed to be changed.

I'm adding @lenaorobei to the thread for more details.

**Comment by mcspronko** [comment]
I can make it backward compatible if this an issue for 2.4. But as far as I am aware, the module-catalog **major** version is going to be increased. Please confirm

**Comment by mcspronko** [comment]
Dear @rogyar 
I've reverted backward-incompatible changes, so you can review the PR.

Appreciate you are looking into it.

**Comment by rogyar** [comment]
Hi @mcspronko. Thanks a lot for the adjustments. Could you take a look at the red static tests, please ("Static tests for B2B" to be precise)? There are some issues that need to be addressed. Some of them are not directly related to your changes but if you have a chance to make it green, it will be awesome. As a rule, the static code analysis issues are pretty minor like missing descriptions or so. 

Please, let me know if you need any help from my end.

**Comment by mcspronko** [comment]
Hi @rogyar 

Please advise what should we do with the failing functional tests, that I believe is somehow random.

Here is the output.
```
no such element: Unable to locate element: {"method":"css selector","selector":".block.widget.block-compared-products-grid"} (Session info: chrome=65.0.3325.162) (Driver info: chromedriver=2.36.540471 (9c759b81a907e70363c6312294d30b6ccccc2752),platform=Linux 3.10.0-693.17.1.el7.x86_64 x86_64) (WARNING: The server did not provide any stacktrace information) Command duration or timeout: 0 milliseconds For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html Build info: version: '3.8.1', revision: '6e95a6684b', time: '2017-12-01T19:05:32.194Z' System info: host: 'extfat.devops.magento.com', ip: '172.17.0.2', os.name: 'Linux', os.arch: 'amd64', os.version: '3.10.0-693.17.1.el7.x86_64', java.version: '1.8.0_181' Driver info: org.openqa.selenium.chrome.ChromeDriver Capabilities {acceptInsecureCerts: false, acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: chrome, chrome: {chromedriverVersion: 2.36.540471 (9c759b81a907e7..., userDataDir: /tmp/.org.chromium.Chromium...}, cssSelectorsEnabled: true, databaseEnabled: false, handlesAlerts: true, hasTouchScreen: false, javascriptEnabled: true, locationContextEnabled: true, mobileEmulationEnabled: false, nativeEvents: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: LINUX, platformName: LINUX, rotatable: false, setWindowRect: true, takesHeapSnapshot: true, takesScreenshot: true, unexpectedAlertBehaviour: , unhandledPromptBehavior: , version: 65.0.3325.162, webStorageEnabled: true} Session ID: f8cfb886013fb072a64cf2caec184630 *** Element info: {Using=css selector, value=.block.widget.block-compared-products-grid}
```

Also, the Semantic Version Checker is red, due to the removal of the optional `registry` dependency from the ProductList helper class.

**Comment by rogyar** [comment]
@magento run all tests

**Comment by rogyar** [comment]
I see the same functional test failing for other PRs. So, it's not related to the current change somehow.

**Comment by rogyar** [comment]
Since you have didn't introduce a new functionality, there's no necessity for creating functional/integration tests. However, if you could cover the newly introduced classes by unit tests, it would be awesome (you will get extra points for that as well). 

Thank you!

**Comment by mcspronko** [comment]
@rogyar, please review the latest changes.

File: app/code/Magento/Catalog/Test/Unit/Helper/Product/ProductListTest.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+{
+    /**
+     * @var ProductList
+     */
+    private $object;
+
+    /**
+     * @var ScopeConfigInterface|MockObject
+     */
+    private $scopeConfig;
```
Comment (rogyar) [inline app/code/Magento/Catalog/Test/Unit/Helper/Product/ProductListTest.php]: Currently, we are advocating suffixing mocks with the `Mock` keyword. So, it's easier to differentiate mocks and real objects through the test. It's not a requirement, but if you could follow this practice, I would appreciate it.

```php
private $scopeConfigMock;
private $toolbarConfigMock;
```

Comment (mcspronko) [reply app/code/Magento/Catalog/Test/Unit/Helper/Product/ProductListTest.php]: @rogyar no problem at all. Can you please refer to documentation to read more about the `Mock` suffix?

Thank you in advance.

Comment (rogyar) [reply app/code/Magento/Catalog/Test/Unit/Helper/Product/ProductListTest.php]: Unfortunately, there's no documentation that explicitly describes the best practices for creating unit tests in DevDocs so far. But we are working on the topic on that matter.

**Comment by rogyar** [comment]
Please, merge the mainline into your current branch to resolve the conflicts. 

Thank you!

**Comment by rogyar** [comment]
Hi @mcspronko. I was going to add a minor change in order to make the static tests green for the automated tests but for some reason, I cannot commit directly to your repo. Could you add the following annotation in the DocBlock of the TollbarTest class, please?

```php
@SuppressWarnings(PHPMD.CouplingBetweenObjects)
```

Thank you

**Comment by mcspronko** [comment]
@rogyar can you please review the PR?
...and maybe merge? :)

**Review by rogyar — DISMISSED** [review]
The failing functional tests are not related to the current PR. 

The failing semantic version checker is failing because the registry decoupling. But such changes are supposed to have some backward-incompatible modifications.

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-7519](https://jira.corp.magento.com/browse/ENGCOM-7519) has been created to process this Pull Request

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @mcspronko . Could you resolve merge conflicts? 
Thanks!

**Comment by mcspronko** [comment]
@engcom-Alfa 

> Hi @mcspronko . Could you resolve merge conflicts?
> Thanks!

Merge conflicts have been resolved.
Please review.

File: app/code/Magento/Catalog/Model/Category/Toolbar/Config.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        $this->currentCategory = $currentCategory;
+        $this->categoryConfig = $categoryConfig;
+    }
+
+    /**
+     * Returns a category default_sort_by attribute, or default sort field from the configuration.
+     *
+     * @return string
+     */
+    public function getOrderField(): string
```
Comment (sshymko) [inline app/code/Magento/Catalog/Model/Category/Toolbar/Config.php:44]: I was thinking how to improve this implementation to be simpler and easier to follow. Here's an equivalent version that avoids duplication of the default behavior.

```php
        try {
            $category = $this->currentCategory->get();
            if ($category->getDefaultSortBy()) {
                return $category->getDefaultSortBy();
            }
        } catch (NoSuchEntityException $exception) {
            // Fall back to the default configuration
        }
        return $this->categoryConfig->getDefaultSortField();
```

I think, in most cases it's better to keep the "positive" logic inside `try..catch` as it guarantees that all its preconditions are met (variable `$category` is defined in this case). Such code is also easier to follow as the sequential logic is grouped together.

**Comment by mcspronko** [comment]
@engcom-Alfa
Any update on this pull request? It's like ages to hear back any feedback and get a positive "approved".

Thank you in advance.

**Comment by mcspronko** [comment]
cc: @rogyar

**Comment by rogyar** [comment]
Hi @mcspronko. Could you merge the latest 2.4-develop changes to your branch and run the tests, please? 

Thank you!

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @mcspronko,
Could you merge recent changes and run the tests again or give us permissions to do so? Unfortunately we don't have permissions for doing that instead of you.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @mcspronko,
Could you review and fix failing tests?

**Comment by mcspronko** [comment]
@ihor-sviziev I opened tests and I don't see what can be fixed.
The SemVer is not something I decide as part of this PR.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@mcspronko could you merge recent changes from 2.4-develop branch?

File: app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php [hunk 3]
```diff
@@ -76,15 +78,20 @@ class ToolbarTest extends TestCase
      */
     protected $pagerBlock;
 
-    protected function setUp(): void
+    /**
+     * @var ToolbarConfig|MockObject
+     */
+    private $toolbarConfig;
+
+    protected function setUp()
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php:86]: Following change broke the tests. Please fix it like this: 
```suggestion
    protected function setUp(): void
```

File: app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php [hunk 4]
```diff
@@ -173,11 +192,11 @@ public function testGetPagerEncodedUrl()
 
         $this->urlBuilder->expects($this->once())
             ->method('getUrl')
-            ->willReturn($url);
+            ->will($this->returnValue($url));
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php:195]: Please revert such changes in all places

File: app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php [hunk 5]
```diff
@@ -96,13 +103,18 @@ protected function setUp(): void
...
+                'setShowAmounts',
+                'setFrameLength',
+                'setJump',
+                'setLimit',
+                'setCollection',
+                'toHtml'
+            ]);
         $this->urlBuilder = $this->createPartialMock(Url::class, ['getUrl']);
-        $this->scopeConfig = $this->getMockForAbstractClass(ScopeConfigInterface::class);
+        $this->scopeConfig = $this->createMock(ScopeConfigInterface::class);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php:117]: Is there any reason for this change?

File: app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php [hunk 6]
```diff
@@ -7,19 +7,21 @@
 
 namespace Magento\Catalog\Test\Unit\Block\Product\ProductList;
 
+use PHPUnit\Framework\TestCase;
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php:10]: Seems like earlier all imports were sorter, while now they're not. Could you run optimize imports in phpstorm for all changes files and commit?

File: app/code/Magento/Catalog/Helper/Product/ProductList.php [hunk 7]
```diff
@@ -45,15 +38,21 @@ class ProductList
...
+    /**
+     * ProductList constructor.
      * @param ScopeConfigInterface $scopeConfig
-     * @param Registry $coreRegistry
+     * @param ToolbarConfig $toolbarConfig
      */
     public function __construct(
         ScopeConfigInterface $scopeConfig,
-        Registry $coreRegistry = null
+        ToolbarConfig $toolbarConfig
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Helper/Product/ProductList.php:52]: This class marked as `@api`, so we can't just replace existing argument with a new one. Please introduce new dependecy as optional argument

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by slavvka** [comment]
@magento run all tests

**Comment by slavvka** [comment]
Unit tests failure:
```
06:03:22 PHP Fatal error:  Declaration of Magento\Catalog\Test\Unit\Block\Product\ProductList\ToolbarTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /var/www/html/app/code/Magento/Catalog/Test/Unit/Block/Product/ProductList/ToolbarTest.php on line 86
```

**Comment by ihor-sviziev** [comment]
Hi @mcspronko,
Will you be able to update this PR?

**Comment by sidolov** [comment]
@mcspronko  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**